src/js/stores/document.js

```diff
@@ -166,18 +173,37 @@ define(function (require, exports, module) {
             if (suppressChange) {
                 return;
             }
-
-            // If some selected layer remains uninitialized, a new document will
-            // come along shortly. Wait until then to trigger the change event.
-            var initialized = nextDocument.layers && nextDocument.layers.selected
-                .every(function (layer) {
-                    return layer.initialized;
-                });
-
-            if (initialized) {
-                // Include the document in the event payload
-                this.emit(changeEventName || "change", nextDocument);
+            
+            var promise = Promise.bind(this);
+
+            // Notify change of layer selection state.
+            if (oldDocument) {
+                promise = promise.then(function () {
+                    oldDocument.layers.layers.forEach(function (layer) {
```

Comment:
I agree, and we should definitely optimize for the select-all operation: it is extremely slow to select all layers in a large document.